docs/expression.md

```diff
@@ -0,0 +1,221 @@
+
+Expression: (a + b) * c + 7
+
+![Expression tree of the expression (a+b)*c+7](https://upload.wikimedia.org/wikipedia/commons/thumb/9/98/Exp-tree-ex-11.svg/500px-Exp-tree-ex-11.svg.png)
+
+<sub>Source: [wikimedia](https://commons.wikimedia.org/wiki/File:Exp-tree-ex-11.svg)</sub>
+
+For more information on how expression trees work and are build see the [wikipedia article](https://en.wikipedia.org/wiki/Binary_expression_tree) on binary expression tree's.
+
+__EBNF Notation__
+```ebnf
+binary_expression = expression, operator, expression;
+```
+
+<sub>see: [operators](operators.md)</sub>
+
+__Example__
+```ttr
+var a
+
+a = 2 * 3 
+
+// 2 * 3 
+// is a binary expression of 2 terms with an multiply operation.
+
+var b
+
+b = a / 1
+
+// a / 1 
+// is a binary expression of 2 terms with an division operation.
+
+val c = a * b + 12 / 2
+
+// a * b + 12 / 2 
+// is a binary expression of two binary expressions. When resolving these it will result as follows
+// binary expression left: a * b (Which is a also binary expressions)
+// operation: Add (+)
+// binary expression right 12 / 2 (Which is a also binary expressions)
+
+// The end result of c = 42
+```
+
+## Unary expression  
+
+Used for applying a single operator to an expression. For instance: using a minus operator to invert the state of an mathematical expression.
+
+__EBNF Notation__
+```ebnf
+unary_expression = minus, expression;
+```
+
+__Example__
+```ttr
+var a = 12
+a = - 12
+
+// - 12
+// is a unary expression that inverts the state of literal to a negative number
+```
+
+## Grouping expression
+
+A grouping is used for telling the interpreter to resolve the grouped expression first. This is used to resolve an nested expression with an higher order of operation before one with an lower one. Using the grouping expression to change the order of operations also changes the outcome. The classic example is [PEMDAS](https://pemdas.info) for basic mathematical expressions.
+
+__EBNF Notation__
+```ebnf
+grouping_expression = "(", expression, ")";
+```
+
+__Example__
+```ttr
+// Using the normal order of operations
+// the 5 * 2 is resolves first and then it will add the 1.
+
+1 + 5 * 2
+
+// Results in 11
+
+// Using the grouping expression we force
+// the interpreter to resolve the 1 + 5 first
+// and multiply with 2 afterwards.
+
+(1 + 5) * 2
+
+// Results in 12
+```
+
+## Assignment expression
+
+Used for binding a result of an expression to a variable. This can be used in deceleration and as expression. It consists of the name of the variable that will bind the expression result. The type of assignment. For instance: a basic assignment that takes the expression results and assigns it to the variable; addition assignment for adding the result of the expression to the existing value of the variable or division assignment that divides the existing value of the variable with the result of the expression.
```

Comment:
```suggestion
Used for binding a result of an expression to a variable. This can be used in declarations and as an expression. It consists of the name of the variable that will bind the expression result, the type of assignment and a expression to evaluate. The expression is evaluated and assigned to the variable using the given operator type.
```